### PR TITLE
drivers/ieee802154: Fix coverity #178236, #178237 Out of bounds access

### LINKED ISSUE #4591: [CID: 178237] memory corruption in drivers/ieee802154/ieee802154_mcr20a.c
Static code scan issues seen in File: /drivers/ieee802154/ieee802154_mcr20a.c
 Category: Memory - corruptions
 Function: _mcr20a_write_burst
 Component: Drivers
 Please fix or provide comments to square it off in coverity in the link: https://scan9.coverity.com/reports.htm#v32951/p12996

### LINKED ISSUE #4592: Coverity issue seen with CID: 178236
Static code scan issues seen in File: /drivers/ieee802154/ieee802154_mcr20a.c
 Category: Memory - corruptions
 Function: _mcr20a_read_burst
 Component: Drivers
 Please fix or provide comments to square it off in coverity in the link: https://scan9.coverity.com/reports.htm#v32951/p12996

### PR DESCRIPTION
**Pull request by dleach02**
Fixed the exposure to an out of bounds access of data structure due to
improper size test ("off by one") in two places.

Fixes #4592
Fixes #4591

Signed-off-by: David Leach <david.leach@nxp.com>

### CODE REVIEW

File: drivers/ieee802154/ieee802154_mcr20a.c [hunk 1]
```diff
@@ -202,12 +202,13 @@ bool _mcr20a_write_burst(struct mcr20a_spi *spi, bool dreg, u16_t addr,
 {
 	bool retval;
 
-	k_sem_take(&spi->spi_sem, K_FOREVER);
```
Comment (jukkar) [inline drivers/ieee802154/ieee802154_mcr20a.c]: This looks like unrelated change, perhaps this could be placed in another commit.

File: drivers/ieee802154/ieee802154_mcr20a.c [hunk 2]
```diff
@@ -202,12 +202,13 @@ bool _mcr20a_write_burst(struct mcr20a_spi *spi, bool dreg, u16_t addr,
 {
```
Comment (jukkar) [inline drivers/ieee802154/ieee802154_mcr20a.c]: The commit message subject mentions coverity ids, these should be placed in body.
So subject could say "drivers/ieee802154: Fix out of bounds access"
and in body you can mention the coverity numbers

Coverity-CID: 178236
Coverity-CID: 178237

**Review by jukkar — CHANGES_REQUESTED** [review]

File: drivers/ieee802154/ieee802154_mcr20a.c (hunk 2 above)
Comment (tbursztyka) [reply drivers/ieee802154/ieee802154_mcr20a.c]: Title should be: "drivers/ieee802154: Fix out of bounds access in mcr20a driver"

File: drivers/ieee802154/ieee802154_mcr20a.c [hunk 3]
```diff
@@ -202,12 +202,13 @@ bool _mcr20a_write_burst(struct mcr20a_spi *spi, bool dreg, u16_t addr,
...
 	bool retval;
 
-	k_sem_take(&spi->spi_sem, K_FOREVER);
-
-	if ((len + 2) > sizeof(spi->cmd_buf)) {
+	if ((len + 2) >= sizeof(spi->cmd_buf)) {
 		SYS_LOG_ERR("Buffer length too large");
```
Comment (tbursztyka) [inline drivers/ieee802154/ieee802154_mcr20a.c]: hopefully in a near future we'll be able to get rid of that weird check, using "new" (some months old already..) SPI API.

**Review by tbursztyka — CHANGES_REQUESTED** [review]
commit message should be changed indeed.

File: drivers/ieee802154/ieee802154_mcr20a.c (hunk 2 above)
Comment (jukkar) [reply drivers/ieee802154/ieee802154_mcr20a.c]: Just noticed that there are two bugs about these issues, these should be mentioned in body like this

Fixes #4592 
Fixes #4591

**Review by jfischer-no — CHANGES_REQUESTED** [review]
pls see/review 114d671e24e47fe0c56f319d368ff07642d2e13c and PR #1135

File: drivers/ieee802154/ieee802154_mcr20a.c (hunk 1 above)
Comment (dleach02) [reply drivers/ieee802154/ieee802154_mcr20a.c]: It is related. The coverity bug is a memory overwrite bug. The "off by one" error is fixed by the changing the test but the SYS_LOG_ERR message doesn't prevent the overwrite from happening. Putting a return 0 does prevent the overwrite but requires changing how you deal with the semaphore. I had two choices, 1) either give back the semaphore before returning 0 or 2) moving the "take" to after the size test. I selected option 2.

File: drivers/ieee802154/ieee802154_mcr20a.c (hunk 2 above)
Comment (dleach02) [reply drivers/ieee802154/ieee802154_mcr20a.c]: Okay, those bugs didn't exist when I did this. I did ask though on the reflector if there was an automated process to convert Coverity bugs into GitHub issues... I'm guessing this answers my question, "no".

**Comment by dleach02** [comment]
commit comment changed according to requests.

**Comment by jukkar** [comment]
> commit comment changed according to requests.

I think you need to edit the commit in your patch and then re-submit the patch, or is github somehow merging the PR description and commit message together. If you open your commit message, then that has now different text than the PR description.

File: drivers/ieee802154/ieee802154_mcr20a.c (hunk 1 above)
Comment (jfischer-phytec-iot) [reply drivers/ieee802154/ieee802154_mcr20a.c]: The need for the cmd_buf is in the worst case (dreg is false) len + 2 (one octet for indirect write cmd + one octet for the address). I do not see the issue here, maybe someone can explain it to me.

Comment (dleach02) [reply drivers/ieee802154/ieee802154_mcr20a.c]: This took me a moment to see as well. Coverity is smart enough to know that the test is checking to see that (len+2) > sizeof(cmd_buf) is potentially problematic later in the code because a maximum value of len at this test could be 10. But later in the code here a value of 10 would have a problem of a one byte over write in:

memcpy(&spi->cmd_buf[2], data_buf, len);

**Comment by dleach02** [comment]
Incorrect change that introduced its own "off-by-one". Correct fix already captured in different PR.

**Comment by jukkar** [comment]
> commit comment changed according to requests.

You changed the PR description, the actual commit is still unchanged. To change the commit message, you need to resend the patch. Add also coverity ids (both) to commit message, the syntax is

Coverity-CID: 178236
Coverity-CID: 178237

**Comment by dleach02** [comment]
@jukkar , I'm trying to close this PR. I don't want it merged as this was fixed in a different pull request that has been merged.
